node/src/services/blockchain/handler.rs

```diff
@@ -72,19 +207,351 @@ impl ActorEventLoop<BlockchainService> for BlockchainServiceEventLoop {
+                (),
+                (),
+                (),
+                (),
+            ),
+        );
+
+        // Getting signer public key.
+        let caller_pub_key = self.keystore.sr25519_public_keys(KEY_TYPE).pop().expect(
+            format!(
+                "There should be at least one sr25519 key in the keystore with key type '{:?}' ; qed",
+                KEY_TYPE
+            )
+            .as_str(),
+        );
+
+        // Sign the payload.
+        let signature = raw_payload
+            .using_encoded(|e| self.keystore.sr25519_sign(KEY_TYPE, &caller_pub_key, e))
+            .expect("The payload is always valid and should be possible to sign; qed")
+            .expect("They key type and public key are valid because we just extracted them from the keystore; qed");
+
+        // Construct the extrinsic.
+        UncheckedExtrinsic::new_signed(
+            function.clone(),
+            storage_hub_runtime::Address::Id(<sp_core::sr25519::Public as Into<
+                storage_hub_runtime::AccountId,
+            >>::into(caller_pub_key)),
+            polkadot_primitives::Signature::Sr25519(signature.clone()),
+            extra.clone(),
+        )
+    }
+
+    /// Get an extrinsic from a block.
+    async fn get_extrinsic_from_block(
+        &self,
+        block_hash: H256,
+        extrinsic_hash: H256,
+    ) -> Result<Extrinsic> {
+        // Get the block.
+        let block = self
+            .client
+            .block(block_hash)
+            .expect("Failed to get block. This shouldn't be possible for known existing block hash; qed")
+            .expect("Block returned None for known existing block hash. This shouldn't be the case for a block known to have at least one transaction; qed");
+
+        // Get the extrinsics.
+        let extrinsics = block.block.extrinsics();
+
+        // Find the extrinsic index in the block.
+        let extrinsic_index = extrinsics
+            .iter()
+            .position(|e| {
+                let hash = Blake2Hasher::hash(&e.encode());
+                hash == extrinsic_hash
+            })
+            .expect("Extrinsic not found in block. This shouldn't be possible if we're looking into a block for which we got confirmation that the extrinsic was included; qed");
+
+        // Get the events from storage.
+        let events_in_block = self.get_events_storage_element(block_hash)?;
+
+        // Filter the events for the extrinsic.
+        // Each event record is composed of the `phase`, `event` and `topics` fields.
+        // We are interested in those events whose `phase` is equal to `ApplyExtrinsic` with the index of the extrinsic.
+        // For more information see: https://polkadot.js.org/docs/api/cookbook/blocks/#how-do-i-map-extrinsics-to-their-events
+        let events = events_in_block
+            .into_iter()
+            .filter(|ev| ev.phase == frame_system::Phase::ApplyExtrinsic(extrinsic_index as u32))
+            .collect();
```

Comment:
Wow… what a hassle. 😅 

We should probably index all storage hub related transactions and easily fetch them later instead of doing all this work. We could use this method as a last resort if the transaction we are searching for was not found in our index.

Reply:
We will eventually do that in the indexer, maybe by then we can optimise this to use that, but I guess for now I'd keep it like this.